### ISSUE TITLE
OpenCL internal compiler error on inline device-side functions

### ISSUE THREAD

**Issue by nevion** [RadeonOpenCompute/ROCm#77]
The error: <unknown>:0:0: in function generate void (float addrspace(1)*, i32, i32, i32, i32): unsupported call to function philox\n\nError: Creating the executable failed: Compiling LLVM IRs to exe.

The kernel sourcecode: https://github.com/arrayfire/arrayfire/blob/devel/src/backend/opencl/kernel/random_engine_philox.cl

The build options:
-cl-std=CL2.0  -D dim_t=long  -D T=float -D THREADS=256 -D RAND_DIST=0 -D ELEMENTS_PER_BLOCK=1024

Simply s/inline //g and the error mysteriously disappears.

**Comment by jrprice** [comment]
This sounds like misuse of the `inline` keyword. Clang conforms to C99 for `inline`, which gives different semantics to either GNU C89 or C++. See [here](http://clang.llvm.org/compatibility.html#inline) for more information.

You should just be able to use `static inline` instead of `inline`.

**Comment by nevion** [comment]
@jrprice weird but I don't recall ever reading about static inline with any OpenCL standard.  As such, still a bug.  Interested to try it though to see but I'm afraid still a bug in the end.  Also, since such optimizations for inlining should be enabled by default and can only be explicitly disabled... well, you know - still a bug. I'll let you know if I can compile with ```static inline``` if I find the time.

**Comment by nevion** [comment]
Hey @gstoner are you sure this should be closed? opencv/opencv#8185 has the issue too in addition to my report of it above.  It looks like a compiler error...

**Comment by boxerab** [comment]
Can confirm that this error still happens on rocm 1.6

**Comment by gstoner** [comment]
Remember OpenCL C language is based on C99, so it uses the C99 semantics (which are different than the c89 or c++ semantics) for the inline keyword. For C99 “static inline” is the proper declaration for functions that should be inlined in the current translation unit.

**Comment by nevion** [comment]
@gstoner alright, I'm a believer, but not a fan of this C99 feature - even as an occational language lawyer I had a hard time digesting that and this summary was of great aid  : https://stackoverflow.com/questions/216510/extern-inline/216546#216546

In the C++ kernel language coming, will that as I expect be C++ rules on inlining?  What's the status of C++ kernelside support these days?

Also, I think rocm is the first platform to do the C99 rules on inlining correctly... which means all the code out in the wild is wrong and they don't know it - and will result in compiler errors.  Perhaps this needs some faq entry that google will pick up somewhere?

**Comment by jrprice** [comment]
> To make matters worse, for those unlucky few supporting multiple opencl platforms, this is going to be ugly to support. I'm not sure it's as simple as a macro decorating the front of a function since per my reading one needs to make sure external function definitions are always available...

Just use `static inline` as Greg suggested - this works with all platforms, both those that conform to C99 like ROCm, and those that don't like NVIDIA/Intel etc.

**Comment by nevion** [comment]
@jrprice can one do that in a header file and get the right behavior?

**Comment by jrprice** [comment]
> @jrprice can one do that in a header file and get the right behavior?

If you have the full function definition in the header, the yes, `static inline` should work fine (though I can't say I've tried this). If it's just a declaration in the header with the definition in some other translation unit that you'll later link against, I assume you want `extern inline` on the definition.

**Comment by gstoner** [comment]
This is opencl spec.

Get Outlook for iOS<https://aka.ms/o0ukef>



On Sat, Jul 8, 2017 at 9:49 AM -0500, "James Price" <notifications@github.com<mailto:notifications@github.com>> wrote:


To make matters worse, for those unlucky few supporting multiple opencl platforms, this is going to be ugly to support. I'm not sure it's as simple as a macro decorating the front of a function since per my reading one needs to make sure external function definitions are always available...

Just use static inline as Greg suggested - this works with all platforms, both those that conform to C99 like ROCm, and those that don't like NVIDIA/Intel etc.

**Comment by nevion** [comment]
@gstoner please read my previous 2 posts again - and that out of date excerpt skips over a real problem - 3 good opencl developers have encountered this obtuse c99 rule and have violated the standard in production code with lots of headscratching... to that, it's probably a good idea for a faq entry since ROCm is the only platform who does it correctly (resulting in a bizarre error) and you're trying to gain marketshare.  I dont't think a closed github issue will show up on google well.  I'm not calling out ROCm for not being standards compliant, I'm trying trying to make sure a pitfall and an entrenched incorrect technique of inlining is addressed.  And static inline is not quite the same in some unsavory ways. Also, there's the unanswered question I asked too regarding status of kernel side OpenCL C++.

Also, I commend your high availability using email to reply, and on the weekend - and I see just how much you're doing it these days but unfortunately for you github culture allows for editing of posts and I do that alot, prior to a reply.  I don't want to send you even more posts for corrections and amendments nor to make issue threads much longer with successive posts. I'm willing to adopt the arguably bad habit of edit posts or notification to accommodate you but I'd just recommend you reply via github so you see whatever edits  while you craft your reply.  If I could get it all right the first go around, I would... and I do take time crafting my replies.

**Comment by gstoner** [comment]
@nevion   OpenCL C++,  is an area we do not see much market demand.  We now have alternative solutions for C++ with HCC and HIP.   OpenCL 2.0 is just getting to point it broad enough market support and it barely.   If you want Mac OS X, Windows and Linux you're still stuck at OpenCL 1.2. 

The rest of your email, I am not going to even comment on.  Please stay constructive in your conversations.  Remember it the weekend, honestly, I am trying to take a vacation was traveling today.

**Comment by nevion** [comment]
@gstoner that's unfortunate...  not the place for the full debate of C++ demand but I assure you many people are interested beyond the creators.  The developers I know in meat space just want OpenCL and aren't interested in deviating, and they want something composable as per discussion about CUB.  HIP is looked at like a crutch that cannot be trusted like CUDA or OpenCL when we're talking about projects with money behind them, at least for now - said people are nervous about targeting HIP.   HCC per my understanding of it is that it did not allow finer control of memory buffer characteristics and in general is less under the hood than OCL. Time may change that but the sentiment everywhere I've worked is "if we're not using CUDA, we just want OpenCL." - no other solution is desired, that's the word at least from a footsoldier.  I hope you guys revisit the decision in the future - I think if you build it, they will come.  And the other builders will deliver as well.

Also,  I really do hope your weekend and vacation is salvaged.  If not, disable your outlook account until Monday :--).  See I didn't send an email - I posted on github issue threads, and edited the some hours prior to your reply. You missed edits several times over the year when you reply some time later (perhaps just when you're using your phone?) and I'm seeking a way of making sure it doesn't happen needlessly.  Consider using [github's android app](https://github.com/blog/1187-github-android-app-released) or an [IOS app](https://ioctocat.com/) to view and maybe submit posts rather than outlook if that's the source of the issue.  If that isn't possible or reasonable - I will operate under the assumption that github is email based for you at the cost of more posts.  To me, this is being constructive.
